File: src/Subsets.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Created by kaijiezhou on 8/31/15.
 */
public class Subsets {
    public List<List<Integer>> subsets(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> result=new ArrayList<>();
        result.add(new ArrayList<Integer>());
        for(int i=0;i<nums.length;i++){
            List<List<Integer>> copy=new ArrayList<>(result);
            for(List<Integer> cur:copy){
                List<Integer> elem=new ArrayList<>(cur);
                if(elem.size()==0||elem.get(elem.size()-1)<nums[i]){
                    elem.add(nums[i]);
                    result.add(elem);
                }
            }
        }
        return result;
    }
    public static void main(String[] args) {
        int[]nums={1,2,3};
        System.out.println(new Subsets().subsets(nums));
    }
}

File: python/subsets.py
__author__ = 'kaijiezhou'

class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        results=[]
        results.append([])
        for i in xrange(0, len(nums)):
            snapshot=list(results)
            for result in snapshot:
                elem=list(result)
                if len(elem)==0 or elem[len(elem)-1]< nums[i]:
                    elem.append(nums[i])
                    results.append(elem)
        return results
